src/MIDebugEngine/Engine.Impl/ExceptionManager.cs

```diff
@@ -425,22 +452,30 @@ private async Task FlushSettingsUpdates()
             }
         }
 
-        private async Task UpdateCatagory(Guid categoryId, ExceptionCategorySettings categorySettings, SettingsUpdates updates)
+        private async Task UpdateCategory(Guid categoryId, ExceptionCategorySettings categorySettings, SettingsUpdates updates)
         {
             // Update the category
-            if (updates.NewCategoryState.HasValue && (
-                updates.NewCategoryState.Value != ExceptionBreakpointStates.None || // send down a rule if the category isn't in the default state
-                categorySettings.CurrentRules.Count != 0)) // Or if we have other rules for the category that we need to blow away
+            lock (categorySettings.CurrentRules)
             {
-                ExceptionBreakpointStates newCategoryState = updates.NewCategoryState.Value;
-                categorySettings.CategoryState = newCategoryState;
-                categorySettings.CurrentRules.Clear();
-
-                IEnumerable<ulong> breakpointIds = await _commandFactory.SetExceptionBreakpoints(categoryId, null, newCategoryState);
-                if (newCategoryState != ExceptionBreakpointStates.None)
+                if (updates.NewCategoryState.HasValue && (
+                    updates.NewCategoryState.Value != ExceptionBreakpointStates.None || // send down a rule if the category isn't in the default state
+                    categorySettings.CurrentRules.Count != 0)) // Or if we have other rules for the category that we need to blow away
                 {
-                    ulong breakpointId = breakpointIds.Single();
-                    categorySettings.CurrentRules.Add("*", breakpointId);
+                    ExceptionBreakpointStates newCategoryState = updates.NewCategoryState.Value;
+                    categorySettings.CategoryState = newCategoryState;
+
+                    // remove exception breakpoints before categorySettings.CurrentRules is cleared
+                    _commandFactory.RemoveExceptionBreakpoint(categoryId, categorySettings.CurrentRules.Values);
+
+                    categorySettings.CurrentRules.Clear();
+
+                    // only do a generic catch throw if C++ exceptions category is checked
+                    if (newCategoryState != ExceptionBreakpointStates.None)
+                    {
+                        IEnumerable<ulong> breakpointIds = _commandFactory.SetExceptionBreakpoints(categoryId, null, newCategoryState).Result;
```

Comment:
It is bad practice to use `.Result` which causes it to be synchronous. 

```suggestion
                        await IEnumerable<ulong> breakpointIds = _commandFactory.SetExceptionBreakpoints(categoryId, null, newCategoryState);
```